lib/CGameInterface.cpp

```diff
@@ -22,256 +37,245 @@
  *
  */
 
-#ifdef VCMI_ANDROID
-// we can't use shared libraries on Android so here's a hack
-extern "C" DLL_EXPORT void VCAI_GetAiName(char* name);
-extern "C" DLL_EXPORT void VCAI_GetNewAI(std::shared_ptr<CGlobalAI> &out);
-
-extern "C" DLL_EXPORT void StupidAI_GetAiName(char* name);
-extern "C" DLL_EXPORT void StupidAI_GetNewBattleAI(std::shared_ptr<CGlobalAI> &out);
-
-extern "C" DLL_EXPORT void BattleAI_GetAiName(char* name);
-extern "C" DLL_EXPORT void BattleAI_GetNewBattleAI(std::shared_ptr<CBattleGameInterface> &out);
-#endif
-
 template<typename rett>
-std::shared_ptr<rett> createAny(const boost::filesystem::path& libpath, const std::string& methodName)
+std::shared_ptr<rett> createAny(const boost::filesystem::path &libpath, const std::string &methodName)
 {
-	typedef void(*TGetAIFun)(std::shared_ptr<rett>&);
-	typedef void(*TGetNameFun)(char*);
+    typedef void(*TGetAIFun)(std::shared_ptr<rett>&);
+    typedef void(*TGetNameFun)(char*);
 
-	char temp[150];
+    char temp[150];
 
-	TGetAIFun getAI = nullptr;
-	TGetNameFun getName = nullptr;
+    TGetAIFun getAI = nullptr;
+    TGetNameFun getName = nullptr;
 
 #ifdef VCMI_ANDROID
-	// this is awful but it seems using shared libraries on some devices is even worse
-	const std::string filename = libpath.filename().string();
-	if (filename == "libVCAI.so")
-	{
-		getName = (TGetNameFun)VCAI_GetAiName;
-		getAI = (TGetAIFun)VCAI_GetNewAI;
-	}
-	else if (filename == "libStupidAI.so")
-	{
-		getName = (TGetNameFun)StupidAI_GetAiName;
-		getAI = (TGetAIFun)StupidAI_GetNewBattleAI;
-	}
-	else if (filename == "libBattleAI.so")
-	{
-		getName = (TGetNameFun)BattleAI_GetAiName;
-		getAI = (TGetAIFun)BattleAI_GetNewBattleAI;
-	}
-	else
-		throw std::runtime_error("Don't know what to do with " + libpath.string() + " and method " + methodName);
+#if THIS_COULD_BE_USED_WHEN_ANDROID_FIXES_RTTI_SUPPORT
+    logGlobal->debugStream() << "Opening AI from " << libpath.string();
+    void *dll = dlopen(libpath.string().c_str(), RTLD_LOCAL | RTLD_LAZY);
+    if (dll)
+    {
+        getName = (TGetNameFun) dlsym(dll, "GetAiName");
+        getAI = (TGetAIFun) dlsym(dll, methodName.c_str());
+    }
+    else
+        logGlobal->errorStream() << "Error: " << dlerror();

```

Comment:
Whoops, true.
I think I should remove that whole "fixed rtti for android" block. I don't think it can be used in near future and it would require to change much more in the build process anyway.